chore(xunit): refactor xUnit reporter class (#5219)
cocotb/cocotb: docs/source/junit.rst

.. _junit:

******************
JUnit Tests Report
******************

Generated XML file with tests results by cocotb regression manager is compatible with `Jenkins xUnit schema version 2`_.

Only few XML elements and attributes are used from above XML schema when generating XML tests report
that are commonly supported in other available JUnit/xUnit XML schemas.

> This allow to support wide range of different CI environments like GitHub Actions, GitLab CI, Jenkins CI and others.

Author: > Does GHA even parse the junit xml file itself? Or only with Action?

@imphil Unfortunately not. You must use one of existing actions like [junit-report-action](https://github.com/marketplace/actions/junit-report-action). I should add this information as well :thinking: I could add an example to these three CI environments. But generally there are dozens of CI environments.

.. _junit-attachments:

Attachments
===========

Simulation log and waveform files will be attached to generated JUnit XML file by using :ref:`junit-attributes-property` and
:ref:`junit-attributes-system-out` XML elements. List of files that should be attached can be defined by using

> the :envvar:`COCOTB_RESULTS_ATTACHMENTS` environment variable.

Reviewer: FFR we split lines based on sentences (or if the line gets really long). This reduces diffs in the future.
```suggestion
Simulation log and waveform files will be attached to generated JUnit XML file by using :ref:`junit-attributes-property` and :ref:`junit-attributes-system-out` XML elements.
The :envvar:`COCOTB_RESULTS_ATTACHMENTS` environment variable is used to define the list of attachments.
```

> .. note::

Reviewer: ```suggestion
pytest Plugin
-------------
```

    To include file attachments in :ref:`junit-attributes-system-out` XML element when generating JUnit XML file from pytest,
    set the pytest ``junit_logging`` option to ``system-out`` or ``all``:

    .. code:: shell

        pytest --override-ini=junit_logging=system-out --junit-xml=junit.xml ...

Example:

.. code:: xml

    <properties>
      <property name="attachment" value="sim_build/dff.vcd" />
      <property name="attachment" value="sim_build/dff.log" />
    </properties>
    <system-out>[[ATTACHMENT|sim_build/dff.vcd]]
    [[ATTACHMENT|sim_build/dff.log]]
    </system-out>

.. _junit-paths:

Paths
=====

The :envvar:`COCOTB_RESULTS_RELATIVE_TO` environment variable can be used to convert all absolute paths reported in
generated XML file to relative ones including XML attributes, properties and text.

Example:

.. code:: shell

    COCOTB_RESULTS_RELATIVE_TO="${CI_PROJECT_DIR:-$(pwd)}" pytest|make ...

> .. note::

Reviewer: ```suggestion
pytest Plugin
-------------
```

    When using the :ref:`pytest-support`, the path hint used to convert absolute paths to relative ones is determined
    from where the ``pytest`` command was invoked. Use the :envvar:`COCOTB_RESULTS_RELATIVE_TO` environment variable to
    override that.

.. _junit-attributes:

Attributes
==========

> Only subset of XML elements and attributes from `Jenkins xUnit schema version 2`_ are supported.

Reviewer: ```suggestion
Only a subset of XML elements and attributes from `Jenkins xUnit schema version 2`_ are supported.
```

    </testsuite>

.. _junit-attributes-testcase:

testcase
--------

Reported test.

It may contain ``properties``, ``skipped``, ``failure``, ``error``, ``system-out`` and ``system-err`` XML elements.

List of supported XML attributes:

* ``classname`` - Python module where reported test was defined.
* ``name`` - name of reported test, mostly name of Python function or Python class.
* ``time`` - total wall time of executed test in seconds.

Example:

.. code:: xml

    <testcase classname="test_dff" name="dff_simple_test" time="0.001">
    </testcase>

.. _junit-attributes-properties:

properties
----------

List of ``property`` XML elements.

No XML attributes.

.. code:: xml

    <properties>
      <property name="cocotb" value="True" />
    </properties>

.. _junit-attributes-property:

property
--------

Reported data by test that cannot be represtented as XML attribute in ``testcase`` XML element.

List of supported XML attributes:

* ``name`` - name of test property.
* ``value`` - value of test property.

List of supported properties:

> * ``cocotb`` - mark ``testcase`` as cocotb test. If defined, default value is ``True``.

Author: @imphil I can re-phrase that or just remove it. If you have multiple JUnit XML tests reports, generated from C/C++ software, Python and Cocotb, it is a good idea to have some kind of identifier after merging all tests reports into single tests report (this is often happening automatically on CI side) to distinguish them.

Makefiles and cocotb Runner
---------------------------

Use the :envvar:`COCOTB_RESULTS_FILE` environment variable to specify the output file name when invoking ``make`` or the cocotb Runner.

.. code:: shell

    COCOTB_RESULTS_FILE=junit.xml make

pytest Plugin
-------------

Use the ``--junit-xml`` command line option to specify the output file name when invoking ``pytest``.

.. code:: shell

    pytest --junit-xml=junit.xml

.. _junit-attachments:

Adding Attachments
==================

It is possible to include file attachments like simulation logs or waveform files in the generated results file.
The file contents are not included directly in the results file,
but are included as file links which certain CI tools like Jenkins xUnit plugin will interpret, upload, and provide to the user as hyperlinks to the uploaded files.

Makefiles and cocotb Runner
---------------------------

The :envvar:`COCOTB_RESULTS_ATTACHMENTS` environment variable is a comma-separated list of files to attach.

.. code:: shell

    COCOTB_RESULTS_ATTACHMENTS=sim.log,sim_build/waveform.vcd make

pytest Plugin
-------------

> Set the ``junit_logging`` option to ``system-out`` or ``all``.

Reviewer 2: ...to do what? Is this a prerequisite for being able to attach files?

Reviewer: I actually have no idea. I thought the same thing.

Reviewer: ```suggestion
Setting the ``junit_logging`` pytest INI config option to ``system-out`` or ``all`` is required for attachments to be added when using the pytest plugin.
```

.. TODO Why does this work?

.. code:: shell

    pytest --override-ini=junit_logging=system-out --junit-xml=junit.xml ...

.. _junit-paths:

Outputting Relative Paths
=========================

> The :envvar:`COCOTB_RESULTS_RELATIVE_TO` environment variable can be used to convert all absolute paths reported in the generated XML file to relative ones including XML attributes, properties and text.

Reviewer 2: ```suggestion
The :envvar:`COCOTB_RESULTS_RELATIVE_TO` environment variable can be used to convert all absolute paths reported in the generated XML file to relative ones, including XML attributes, properties and text.
```
?
